tsdb/tx.go

```diff
@@ -248,15 +254,27 @@ func (l *LocalMapper) Open() error {
 	l.txn = txn
 
 	// create a bolt cursor for each unique series id
-	l.cursors = make([]*bolt.Cursor, len(l.seriesKeys))
+	l.cursors = make([]*shardCursor, len(l.seriesKeys))
 
 	for i, key := range l.seriesKeys {
+		// Retrieve key bucket.
 		b := l.txn.Bucket([]byte(key))
-		if b == nil {
+
+		// Ignore if there is no bucket or points in the cache.
+		if b == nil && len(l.shard.cache[key]) == 0 {
 			continue
 		}
 
-		l.cursors[i] = b.Cursor()
+		// Retrieve a copy of the in-cache points for the key.
+		cache := make([][]byte, len(l.shard.cache[key]))
+		copy(cache, l.shard.cache[key])
```

Comment:
It could be a large `memcpy()` but:

1. it's only going to apply to keys that are actually read
2. it's one contiguous read of the pointers -- not the data itself -- so it should be fast
3. it allows us to not lock the cache during queries and we're write heavy so a copy-on-read seems like it'd make more sense than a copy-on-write.